File: app/assets/javascripts/playmat/helpers/PlaymatResizer.js
/* global define:false */
/* global Phaser:false */
define(['playmat/helpers/Layout'], function (calculateLayout) {
  'use strict';

  var PlaymatResizer = function(playmat) {
    this.playmat = playmat;
  };

  PlaymatResizer.prototype.refreshLayout = function() {
    var p = this.playmat;
    var originalSize = [p.w, p.h];

    p.layout = calculateLayout(p.pileSets, p.layoutString, p.game.world.width / p.game.world.height, p.game.menusVisible);

    p.w = p.layout.width * p.game.cardSet.cardWidth;
    p.h = (p.layout.height - 0.1) * p.game.cardSet.cardHeight;

    if(p.w !== originalSize[0] || p.h !== originalSize[1]) {
      this.resize();
    }
  };

  PlaymatResizer.prototype.resize = function() {
    var p = this.playmat;
    var totalHeight = p.game.world.height;
    var menuHeight = 0;
    if(p.game.menusVisible) {
      totalHeight -= 80;
      menuHeight = 40;
    }
    var widthRatio = p.game.world.width / p.w;
    var heightRatio = totalHeight / p.h;

    var newPosition = p.position;
    var newScale = p.scale;

    if(widthRatio < heightRatio) {
      newScale = new Phaser.Point(widthRatio, widthRatio);
      var yOffset = (totalHeight - (p.h * widthRatio)) / 2;
      if(yOffset > 0 || p.y !== menuHeight) {
        newPosition = new Phaser.Point(0, menuHeight /* yOffset */);
      }
    } else {
      newScale = new Phaser.Point(heightRatio, heightRatio);
      var xOffset = (p.game.world.width - (p.w * heightRatio)) / 2;
      if(xOffset > 0 || p.x !== 0) {
        newPosition = new Phaser.Point(xOffset, menuHeight);
      }
    }
    if(p.game.initialized) {
      p.game.add.tween(p.scale).to({x: newScale.x, y: newScale.y}, 500, Phaser.Easing.Quadratic.InOut, true);
      p.game.add.tween(p.position).to({x: newPosition.x, y: newPosition.y}, 500, Phaser.Easing.Quadratic.InOut, true);
    } else {
      p.scale = newScale;
      p.position = newPosition;
    }
  };

  return PlaymatResizer;
});

File: client/src/main/scala/phaser/playmat/layout/PlaymatResizer.scala
package phaser.playmat.layout

import com.definitelyscala.phaser.Easing.Easing
import com.definitelyscala.phaser.Point
import phaser.playmat.Playmat

import scala.scalajs.js

class PlaymatResizer(p: Playmat) {
  private[this] var lastSize = p.phaser.width -> p.phaser.height

  def refreshLayout() = {
    val originalSize = 0.0 -> 0.0

    p.layout = Layout.calculateLayout(p.pileSets, p.layoutString)

    p.w = p.layout._1 * p.phaser.getSettings.cardSet.cardWidth
    p.h = (p.layout._2 - 0.1) * p.phaser.getSettings.cardSet.cardHeight

    if (p.w != originalSize._1 || p.h != originalSize._2) {
      this.resize()
    }
  }

  def resizeIfChanged() = if (lastSize._1 != p.phaser.width || this.lastSize._2 != p.phaser.height) {
    resize()
  }

  def resize() = {
    utils.Logging.info("Playmat resize...")

    val totalWidth = p.game.world.width
    val widthRatio = totalWidth / p.w

    val totalHeight = p.game.world.height
    val heightRatio = totalHeight / p.h

    this.lastSize = totalWidth -> totalHeight

    var newPosition = p.position
    var newScale = p.scale

    if (widthRatio < heightRatio) {
      newScale = new Point(widthRatio, widthRatio)
      val yOffset = (totalHeight - (p.h * widthRatio)) / 2
      if (yOffset > 0 || p.y != 0) {
        newPosition = new Point(0, 0 /* yOffset */ )
      }
    } else {
      newScale = new Point(heightRatio, heightRatio)
      val xOffset = (p.game.world.width - (p.w * heightRatio)) / 2
      if (xOffset > 0 || p.x != 0) {
        newPosition = new Point(xOffset, 0)
      }
    }

    if (p.phaser.initialized) {
      val scaleProps = js.Dynamic.literal("x" -> newScale.x, "y" -> newScale.y)
      p.game.add.tween(p.scale).to(scaleProps, 500, Easing.Default, autoStart = true, delay = 0.0, repeat = 0.0, yoyo = false)

      val posProps = js.Dynamic.literal("x" -> newPosition.x, "y" -> newPosition.y)
      p.game.add.tween(p.position).to(posProps, 500, Easing.Default, autoStart = true, delay = 0.0, repeat = 0.0, yoyo = false)
    } else {
      p.scale = newScale
      p.position = newPosition
    }
  }
}
